netket/hilbert/random/base.py

```diff
@@ -40,210 +47,86 @@ def random_state(hilb, key, size=None, dtype=np.float32):
+        )
     )
 
 
-def _random_state_scalar_default_impl(hilb, key, dtype, batch_rule):
-    return batch_rule(hilb, key, 1, dtype).reshape(-1)
+@dispatch
+def random_state(hilb, key, size: NoneType, *, dtype):  # noqa: F811
+    return random_state(hilb, key, 1, dtype=dtype)[0]
 
 
-def _random_state_batch_default_impl(hilb, key, size, dtype, scalar_rule):
-    keys = jax.random.split(key, size)
-    res = jax.vmap(scalar_rule, in_axes=(None, 0, None), out_axes=0)(hilb, key, dtype)
-    return res
+@dispatch
+def random_state(hilb, key, size: Dim, *, dtype):  # noqa: F811
+    n = int(np.prod(size))
+    return random_state(hilb, key, n, dtype=dtype).reshape(*size, -1)
 
 
-def register_random_state_impl(clz=None, *, scalar=None, batch=None):
-    """
-    Register an implementation for the function generating random
-    state for the given Hilbert space class.
-
-    The rule can be implemented both as a scalar rule and as a batched
-    rule, but the best performance will be obtained by implementing
-    the batched version.
-
-    The missing rule will be auto-implemented from the over.
-
-    scalar must have signature
-        (hilb, key, dtype) -> vector
-    batch must have signature
-        (hilb, key, size, dtype) -> matrix of states
+def flip_state(hilb, key, state, indices):
+    r"""
+    Given a state `σ` and an index `i`, randomly flips `σ[i]` so that
+    `σ_new[i] ≢ σ[i]`.
 
-    The function will be jit compiled, so make sure to use jax.numpy.
-    Hilbert is passed as a static object.
+    Also accepts batched inputs, where state is a batch and indices is a
+    vector of ints.
 
-    Arguments:
-        clz: The class of the hilbert space
-        scalar: The function computing a single random state
-        batch: the function computing batches of random states
+    Returns:
+        new_state: a state or batch of states, with one site flipped
+        old_vals: a scalar (or vector) of the old values at the flipped sites
     """
-    if scalar is None and batch is None:
-        raise ValueError("You must at least provide a scalar or batch rule.")
-
-    scalar_rule = scalar
-    batch_rule = batch
-
-    if scalar is None:
-        if clz is None:
-            clz = list(batch.__annotations__.items())[0]
-        scalar_rule = partial(_random_state_scalar_default_impl, batch_rule=batch_rule)
-
-    if batch is None:
-        if clz is None:
-            clz = list(scalar.__annotations__.items())[0]
-
-        batch_rule = partial(_random_state_batch_default_impl, scalar_rule=scalar_rule)
-
-    random_state_scalar.register(clz, scalar_rule)
-    random_state_batch.register(clz, batch_rule)
-
-
-##############################
-### flip_state functions ###
-##############################
+    if state.ndim == 1:
+        return flip_state_scalar(hilb, key, state, indices)
+    else:
+        return flip_state_batch(hilb, key, state, indices)
 
 
-@singledispatch
+@dispatch
 def flip_state_scalar(hilb, key, state, indx):
```

Comment:
would it be possible to remove scalar and batch versions and replace them with a dispatch? or is it sci fi?

Reply:
In a sane language yes.
In an insane language like python sometimes.

I did it for `random_state` because one of the inputs specifies the size of the output.
Here we should dispatch on the number of dimensions of `state` which is not doable yet.

Apparently it might be doable in Python 3.10. but that's still off.

I'll try to think alternatives.

Reply:
it's ok, this is still somehow an implementation detail but yeah if python 10 supports that we should go for it...!